# My First Data Science Project
---
In the code below, we will:

- Define a function named add_value(), which takes in as an input a parameter x, and returns x + 30
- Assign the integer 20 to a variable named val
- Assign the result of add_value(val) to a variable named result
- Print result    

In [1]:
def add_value(x):
    return x + 30
val = 20

In [2]:
result = add_value(val)

In [3]:
print(result)

50


In [4]:
#import libraries
import pandas as pd
from csv import reader

In [5]:
pwd #location of saved files


'C:\\Users\\detri\\Documents\\Jupyter Projects'

In [6]:
file_name = pd.read_csv(r'C:\Users\detri\Documents\Jupyter Projects\AppleStore.csv')
file_name.head()

id               track_name  size_bytes currency  price  \
0  284882215                 Facebook   389879808      USD    0.0   
1  389801252                Instagram   113954816      USD    0.0   
2  529479190           Clash of Clans   116476928      USD    0.0   
3  420009108               Temple Run    65921024      USD    0.0   
4  284035177  Pandora - Music & Radio   130242560      USD    0.0   

   rating_count_tot  rating_count_ver  user_rating  user_rating_ver      ver  \
0           2974676               212          3.5              3.5     95.0   
1           2161558              1289          4.5              4.0    10.23   
2           2130805               579          4.5              4.5  9.24.12   
3           1724546              3842          4.5              4.0    1.6.2   
4           1126879              3594          4.0              4.5    8.4.1   

  cont_rating        prime_genre  sup_devices.num  ipadSc_urls.num  lang.num  \
0          4+  Social Networking               37                1        29   
1         12+      Photo & Video               37                0        29   
2          9+              Games               38                5        18   
3          9+              Games               40                5         1   
4         12+              Music               37                4         1   

   vpp_lic  
0        1  
1        1  
2        1  
3        1  
4        1

In [7]:
import pandas as pd

# Create function for opening a data file
# If the filename is in the same folder, the location isn't needed
def open_dataset(file_name='AppleStore.csv', header=True):
    # Create a dataset from csv WITH header included
    df = pd.read_csv(file_name, delimiter=',')
    # User list comprehension to create a list of lists from Dataset rows
    data = [list(row) for row in df.values]
    if header:
        # Insert Column names as first list in list of lists
        data.insert(0, df.columns.to_list())
        return data
    else:
        return data

# Open the default AppleStore.csv file and keep the header
all_data = open_dataset(header=True)
# Set the header data to variable named header
header = all_data[0]
# Set the rest of the data to the variable named apps_data
apps_data = all_data[1:]

print(header)
print("----------")
print(apps_data[0:4])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
----------
[[284882215, 'Facebook', 389879808, 'USD', 0.0, 2974676, 212, 3.5, 3.5, '95.0', '4+', 'Social Networking', 37, 1, 29, 1], [389801252, 'Instagram', 113954816, 'USD', 0.0, 2161558, 1289, 4.5, 4.0, '10.23', '12+', 'Photo & Video', 37, 0, 29, 1], [529479190, 'Clash of Clans', 116476928, 'USD', 0.0, 2130805, 579, 4.5, 4.5, '9.24.12', '9+', 'Games', 38, 5, 18, 1], [420009108, 'Temple Run', 65921024, 'USD', 0.0, 1724546, 3842, 4.5, 4.0, '1.6.2', '9+', 'Games', 40, 5, 1, 1]]


In the code above:
- Data from the AppleStore was opened and read
- The "header" or names of the columns was set to a variable named 'header'
- All of the other information was set to a variable named 'apps_data'
- The first 4 columns of the data was printed
- The data set can be downloaded from https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/

| **Column name** | **Description** |
|---|:---|
| "id" | App ID |
| "track_name" | App Name |
| "size_bytes" | Size (in Bytes) |
| "currency" | Currency Type |
| "price" | Price of App |
| "rating_count_tot" | Total # of ratings (for all versions) |
| "rating_count_ver" | Total # of ratings (for current version) |
| "user_rating" | User Rating Value (for all versions) |
| "user_rating_ver" | User Rating Value (for current version) |
| "ver" | Latest App Version |
| "cont_rating" | Content Age Rating |
| "prime_genre" | Primary Genre of App |
| "sup_devices.num" | Number of supporting devices |
| "ipadSc_urls.num" | Number of screenshots showed for display |
| "lang.num" | Number of supported languages |
| "vpp_lic" | Vpp Device Based Licensing Enabled |

In code below:
- Data from the AppleStore was opened and read
- The "header" or names of the columns was set to a variable named 'header'
- All of the other information was set to a variable named 'apps_data'
---
- Calculated the overall average ratings for all apps
- Calculated the avergage ratings for all and current version free apps
- Calculated the avergage ratings for all and current version paid apps
- Calculated the average price of the paid apps
---
- Calculated the overall app size in bytes, converted to MB
- Calculated the largest and smallest free apps size available
- Calculated the largest and smallest paid apps size available
---
- Calculated the number of apps per asge ratings for all apps
- Calculated the number of apps per asge ratings for all free apps
- Calculated the number of apps per asge ratings for all paid apps


In [8]:
import pandas as pd

# Create function for opening a data file
# If the filename is in the same folder, the location isn't needed
def open_dataset(file_name='AppleStore.csv', header=True):
    # Create a dataset from csv WITH header included
    df = pd.read_csv(file_name, delimiter=',')
    # User list comprehension to create a list of lists from Dataset rows
    data = [list(row) for row in df.values]
    if header:
        # Insert Column names as first list in list of lists
        data.insert(0, df.columns.to_list())
        return data
    else:
        return data

# Open the default AppleStore.csv file and keep the header
all_data = open_dataset(header=True)
# Set the header data to variable named header
header = all_data[0]
# Set the rest of the data to the variable named apps_data
apps_data = all_data[1:]
total_length = len(apps_data)
#print(total_length) #7197

In [9]:
# All the starting lists and variables
all_ratings = 0
free_apps_ratings = []
free_apps_ratings_cur = []
paid_apps_ratings = []
paid_apps_ratings_cur = []
price_of_apps = 0
all_apps_size = 0
free_apps_size = []
paid_apps_size = []
content_ratings = {"4+":0, "9+":0, "12+":0, "17+":0}
free_content_ratings = {"4+":0, "9+":0, "12+":0, "17+":0}
paid_content_ratings = {"4+":0, "9+":0, "12+":0, "17+":0}

In [10]:
# Setting data to lists and variables
for row in apps_data:
    size = float(row[2])
    overall_rating = float(row[7])
    current_rating = float(row[8])
    price = float(row[4])
    c_rating = row[10]
    all_ratings = all_ratings + overall_rating
    all_apps_size = all_apps_size + size
    
    if price == 0.0:
        free_apps_ratings.append(overall_rating)
        free_apps_ratings_cur.append(current_rating)
        free_apps_size.append(size)
        if (c_rating in content_ratings) == True:
            free_content_ratings[c_rating] += 1
        else:
            free_content_ratings[c_rating] = 1
    else:
        paid_apps_ratings.append(overall_rating)
        paid_apps_ratings_cur.append(current_rating)
        price_of_apps = price_of_apps + price
        paid_apps_size.append(size)
        if (c_rating in content_ratings) == True:
            paid_content_ratings[c_rating] += 1
        else:
            paid_content_ratings[c_rating] = 1
            
    if (c_rating in content_ratings) == True:
        content_ratings[c_rating] += 1
    else:
        content_ratings[c_rating] = 1

In [11]:
print("Overall number of apps:", total_length)

avg_overall_rating = all_ratings/total_length
print("Overall average ratings for all apps:", round(avg_overall_rating,2))
print("------------")
avg_overall_free_apps_ratings = sum(free_apps_ratings)/len(free_apps_ratings)
print("Overall free apps average ratings:", round(avg_overall_free_apps_ratings,2))

avg_current_free_apps_ratings = sum(free_apps_ratings_cur)/len(free_apps_ratings_cur)
print("Current version, free apps average ratings:", round(avg_current_free_apps_ratings,2))
print("------------")
avg_overall_paid_apps_ratings = sum(paid_apps_ratings)/len(paid_apps_ratings)
print("Overall paid apps average ratings:", round(avg_overall_paid_apps_ratings,2))

avg_current_paid_apps_ratings = sum(paid_apps_ratings_cur)/len(paid_apps_ratings_cur)
print("Current version, paid apps average ratings:", round(avg_current_paid_apps_ratings,2))

avg_overall_price = price_of_apps/len(paid_apps_ratings)
print("Overall average price, of paid apps: $", round(avg_overall_price,2))
print("------------")

free_or_paid = ""
if avg_overall_free_apps_ratings > avg_overall_paid_apps_ratings:
    free_or_paid = "Free apps"
else:
    free_or_paid = "Paid apps"
    
print(free_or_paid, "have a higher overall rating.")
print("------------")

avg_apps_size = all_apps_size/total_length
print("The overall average app size to download is:", round(avg_apps_size/1000000, 2), "MB")
print("------------")

avg_free_apps_size = sum(free_apps_size)/len(free_apps_size)
print("The average free app size to download is:", round(avg_free_apps_size/1000000, 2), "MB")

free_max = round(max(free_apps_size)/100000, 2)
print("The largest size is:", free_max , "MB, or", round(free_max/1000, 2), "GB")
free_min = round(min(free_apps_size)/100000, 2)
print("The smallest size is:", free_min , "MB")

print("------------")
avg_paid_apps_size = sum(paid_apps_size)/len(paid_apps_size)
print("The average paid app size to download is:", round(avg_paid_apps_size/1000000, 2), "MB")

paid_max = round(max(paid_apps_size)/100000, 2)
print("The largest size is:", paid_max , "MB, or", round(paid_max/1000, 2), "GB")
paid_min = round(min(paid_apps_size)/100000, 2)
print("The smallest size is:", paid_min , "MB")
print("------------")

#print(content_ratings)
print("The age ratings for all apps: 4+", content_ratings["4+"], "/ 9+", content_ratings["9+"], "/ 12+", content_ratings["12+"], "/ 17+", content_ratings["17+"])
#print(free_content_ratings)
print("The age ratings for free apps: 4+", free_content_ratings["4+"], "/ 9+", free_content_ratings["9+"], "/ 12+", free_content_ratings["12+"], "/ 17+", free_content_ratings["17+"])
#print(paid_content_ratings)
print("The age ratings for paid apps: 4+", paid_content_ratings["4+"], "/ 9+", paid_content_ratings["9+"], "/ 12+", paid_content_ratings["12+"], "/ 17+", paid_content_ratings["17+"])


Overall number of apps: 7197
Overall average ratings for all apps: 3.53
------------
Overall free apps average ratings: 3.38
Current version, free apps average ratings: 3.06
------------
Overall paid apps average ratings: 3.72
Current version, paid apps average ratings: 3.5
Overall average price, of paid apps: $ 3.96
------------
Paid apps have a higher overall rating.
------------
The overall average app size to download is: 199.13 MB
------------
The average free app size to download is: 147.94 MB
The largest size is: 31484.21 MB, or 31.48 GB
The smallest size is: 7.67 MB
------------
The average paid app size to download is: 265.25 MB
The largest size is: 40259.7 MB, or 40.26 GB
The smallest size is: 5.9 MB
------------
The age ratings for all apps: 4+ 4433 / 9+ 987 / 12+ 1155 / 17+ 622
The age ratings for free apps: 4+ 2466 / 9+ 438 / 12+ 705 / 17+ 447
The age ratings for paid apps: 4+ 1967 / 9+ 549 / 12+ 450 / 17+ 175


| **Description** | **Number** |
|:---|:---|
|Overall Number of Apps|7197|
|Overall avg. ratings|3.53|
|||
|Overall avg. free app ratings|3.38|
|Current ver. avg. free app ratings|3.06|
|||
|Overall avg. paid app ratings|3.72|
|Current ver. avg. paid app ratings|3.5|
|Overall avg. price|$3.96|
|||
|Overall avg. app size|199.13 MB|
|Avg. free app size|147.94 MB|
|Largest free app size|31.48 GB|
|Smallest free app size|7.67 MB|
|||
|Avg. paid app size|265.25 MB|
|Largest paid app size|40.26 GB|
|Smallest paid app size|5.9 MB|
|||
|**Number of apps per age rating**|
|4+|8870|
|9+|1977|
|12+|2314|
|17+|1244|
|||
|**Number of free apps per age rating**|
|4+|2470|
|9+|441|
|12+|709|
|17+|447|
|||
|**Number of paid apps per age rating**|
|4+|1967|
|9+|549|
|12+|450|
|17+|175|